src/client/dfs/dfs_sys.c

```diff
@@ -0,0 +1,1315 @@
+		dfs_release(obj);
+		D_GOTO(out_free_path, rc = EINVAL);
+	}
+
+	*_obj = obj;
+
+out_free_path:
+	sys_path_free(dfs_sys, &sys_path);
+	return rc;
+}
+
+int
+dfs_sys_close(dfs_obj_t *obj)
+{
+	return dfs_release(obj);
+}
+
+int
+dfs_sys_read(dfs_sys_t *dfs_sys, dfs_obj_t *obj, void *buf, daos_off_t off,
+	     daos_size_t *size, daos_event_t *ev)
+{
+	int		rc;
+	d_iov_t		iov;
+	d_sg_list_t	sgl;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (buf == NULL)
+		return EINVAL;
+	if (size == NULL)
+		return EINVAL;
+
+	d_iov_set(&iov, buf, *size);
+	sgl.sg_nr = 1;
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr_out = 1;
+
+	rc = dfs_read(dfs_sys->dfs, obj, &sgl, off, size, ev);
+	if (rc != 0)
+		*size = -1;
+
+	return rc;
+}
+
+int
+dfs_sys_write(dfs_sys_t *dfs_sys, dfs_obj_t *obj, const void *buf,
+	      daos_off_t off, daos_size_t *size, daos_event_t *ev)
+{
+	int		rc;
+	d_iov_t		iov;
+	d_sg_list_t	sgl;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (buf == NULL)
+		return EINVAL;
+	if (size == NULL)
+		return EINVAL;
+
+	d_iov_set(&iov, (void *)buf, *size);
+	sgl.sg_nr = 1;
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr_out = 1;
+
+	rc = dfs_write(dfs_sys->dfs, obj, &sgl, off, ev);
+	if (rc != 0)
+		*size = -1;
+
+	return rc;
+}
+
+int
+dfs_sys_punch(dfs_sys_t *dfs_sys, const char *file,
+	      daos_off_t offset, daos_off_t len)
+{
+	int		rc;
+	struct sys_path	sys_path;
+	dfs_obj_t	*obj;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (file == NULL)
+		return EINVAL;
+
+	rc = sys_path_parse(dfs_sys, &sys_path, file);
+	if (rc != 0)
+		return rc;
+
+	rc = dfs_open(dfs_sys->dfs, sys_path.parent, sys_path.name,
+		      S_IFREG, O_RDWR, 0, 0, NULL, &obj);
+	if (rc != 0) {
+		D_DEBUG(DB_TRACE, "dfs_open() %s failed: (%d)\n",
+			sys_path.name, rc);
+		D_GOTO(out, rc);
+	}
+
+	rc = dfs_punch(dfs_sys->dfs, obj, offset, len);
+
+	dfs_release(obj);
+out:
+	sys_path_free(dfs_sys, &sys_path);
+	return rc;
+}
+
+int
+dfs_sys_opunch(dfs_sys_t *dfs_sys, dfs_obj_t *obj,
+	       daos_off_t offset, daos_off_t len)
+{
+	if (dfs_sys == NULL)
+		return EINVAL;
+
+	return dfs_punch(dfs_sys->dfs, obj, offset, len);
+}
+
+int
+dfs_sys_remove(dfs_sys_t *dfs_sys, const char *path, bool force,
+	       daos_obj_id_t *oid)
+{
+	return dfs_sys_remove_type(dfs_sys, path, force, 0, oid);
+}
+
+int
+dfs_sys_remove_type(dfs_sys_t *dfs_sys, const char *path, bool force,
+		    mode_t mode, daos_obj_id_t *oid)
+{
+	int		rc;
+	struct sys_path	sys_path;
+	struct stat	stbuf;
+	mode_t		type_mask = mode & S_IFMT;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (path == NULL)
+		return EINVAL;
+
+	rc = sys_path_parse(dfs_sys, &sys_path, path);
+	if (rc != 0)
+		return rc;
+
+	/* Can't delete root */
+	if (sys_path.name == NULL)
+		D_GOTO(out_free_path, rc = EBUSY);
+
+	/* Only check the type if passed in */
+	if (type_mask == 0)
+		goto remove;
+
+	/* Stat the object and make sure it is the correct type */
+	rc = dfs_stat(dfs_sys->dfs, sys_path.parent, sys_path.name, &stbuf);
```

Comment:
There is a race condition here, I've actually got a PR somewhere that allows you to pass in a oid to dfs_remove() which will only remove that object rather than simply telling you what it removed.  Outside the scope of this PR however.

Reply:
Oh, duh. Yeah, something like that would be useful here.
More generally, there are several functions in this file that could benefit from perhaps having some functions added to dfs_internal.h